File: wordsearch.c
int isvalid(int i,int j, int numrows, int numcols)
{
  if (i >= 0 && i < numrows && j >= 0 && j < numcols)
  {
    return 1;
  }
  return 0;
}
int backtrack(char **board, int i , int j, char *word, int **shadowgrid, int found, int numrows, int numcols)
{
  /* check around this square */
  //word = word + 1;
  if (*word == '\0' )
  {
    return 1;
  }
    char temp;// = malloc(sizeof(char) * 4);
    //strcpy(temp, board[i][j])
    temp = board[i][j]    ;
    board[i][j] = ' ';
    int ret = 0;
    /* check i, j+1 */
    if (ret == 0 && isvalid(i,j+1, numrows, numcols) && board[i][j+1] == *word)
    {
      shadowgrid[i][j+1] = -1;
      ret = backtrack(board, i , j + 1, word + 1, shadowgrid, 1, numrows, numcols);
      
    }
    if (ret == 0 && isvalid(i,j-1, numrows, numcols)  && board[i][j-1] == *word)
    {
      shadowgrid[i][j-1] = -1;
      ret = backtrack(board, i , j-1, word + 1, shadowgrid, 1, numrows, numcols);

    }
    if (ret == 0 && isvalid(i-1,j, numrows, numcols)  && board[i-1][j] == *word)
    {
      shadowgrid[i-1][j] = -1;
      ret = backtrack(board, i -1 , j, word + 1, shadowgrid, 1, numrows, numcols);

    }
    if (ret == 0 && isvalid(i+1,j, numrows, numcols) && board[i+1][j] == *word)
    {
      shadowgrid[i+1][j] = -1;
      ret = backtrack(board, i+1 , j, word + 1, shadowgrid, 1, numrows, numcols);

    }
    board[i][j] =  temp;
    //free(temp);
    return ret;
}

bool exist(char** board, int boardSize, int* boardColSize, char * word){
  int i = 0;
  int j = 0;
  int numrows = boardSize;
  int numcols = *boardColSize;
  /* malloc shadow grid as int.
     1. we use 0 as unused
     2. -1 as visited
     3. increment from 0 as we find a match
     */
  int **shadowgrid;
  shadowgrid = malloc(sizeof(int) * 1024);
  for (i = 0; i < numrows; i++)
  {
    shadowgrid[i] = malloc(sizeof(int) * 1024);
   // memset(shadowgrid[i], 0 , sizeof(int) * numcols);
  }
  for (i = 0; i < numrows; i++)
  {
      for (j = 0; j < numcols; j++)
      {
        shadowgrid[i][j] = 0;
      }  
  }
  for (i = 0; i < numrows; i++)
  {
    for (j = 0; j < numcols; j++)
    {
      if ((board[i][j]) != word[0])
      {
        //shadowgrid[i][j] = -1;
        continue;
      }
      else
      {
        shadowgrid[i][j] = -1; 
        int ret = backtrack(board, i, j, word+1, shadowgrid, 1, numrows, numcols);
        if (ret == 1)
        {
          return 1;
        }
        else
        {
              int k = 0;
              int l = 0;
              for (k = 0; k < numrows; k++)
              {
                for (l = 0; l < numcols;l++)
                {
                    shadowgrid[k][l] = 0;
                }      
              }
        }
      }
    }
  }
  return 0;


}

File: udemy-50-problems/wordsearch.py
# Parameters:
#  board: List[List[char]]
#  word: str
# return type: bool

def exist(board, word):
  def clearvisited(visited):
    numrows= len(board)
    numcols = len(board[0])

    for i in range(0, numrows):
      for j in range(0, numcols) :
        visited[i][j] = 0

  def isValid(row, col, visited):
    numrows= len(board)
    numcols = len(board[0])
    if(row >= 0 and row < numrows and col >= 0 and col < numcols and visited[row][col] == 0):
      return True
    else:
      return False

  def recExist(board, row, col, visited, index, word):
    if(index == len(word)):
      return True

    else:
      ret1 = False
      ret2 = False
      ret3 = False
      ret4 = False
      if(isValid(row+1, col,visited) and board[row+1][col] == word[index]):
        visited[row+1][col] = 1
        ret1 = recExist(board, row+1, col, visited, index+1, word)
      if(isValid(row-1, col,visited) and board[row-1][col] == word[index]):
        visited[row-1][col] = 1
        ret2 = recExist(board, row-1, col, visited, index+1, word)
      if(isValid(row, col+1,visited) and board[row][col+1] == word[index]):
        visited[row][col+1] = 1
        ret3 =  recExist(board, row, col+1, visited, index+1, word)
      if(isValid(row, col-1,visited) and board[row][col-1] == word[index]):
        visited[row][col-1] = 1
        ret4 =  recExist(board, row, col-1, visited, index+1, word)
      if(ret1 or ret2 or ret3 or ret4):
        return True
      else:
        return False

  visited = []
  for i in range(0, len(board)) :
    visited.append([])
    for j in range(0, len(board[0])) :
      visited[i].append(0)

  for i in range(0, len(board)) :
    for j in range(0, len(board[0])) :
      if(board[i][j] == word[0]):
        clearvisited(visited)
        visited[i][j] = 1
        ret = recExist(board, i, j, visited, 1, word)
        if(ret == True):
          return True

  return False
board = [['K','I','N','T'],['B','I','N','S'],['G','N','Y','I'],['U','O','E','D'],['D','I','B','V'],['H','I','R','T']]

print(exist(board,"CODE")) 

